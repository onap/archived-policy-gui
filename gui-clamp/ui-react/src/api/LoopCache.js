/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2019 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

export default class LoopCache {
	loopJsonCache;

	constructor(loopJson) {
		this.loopJsonCache=loopJson;
	}

	updateMicroServiceProperties(name, newMsProperties) {
			for (var policy in this.loopJsonCache["microServicePolicies"]) {
				if (this.loopJsonCache["microServicePolicies"][policy]["name"] === name) {
					this.loopJsonCache["microServicePolicies"][policy]["configurationsJson"] = newMsProperties;
				}
			}
	}

	updateMicroServicePdpGroup(name, pdpGroup, pdpSubgroup) {
			for (var policy in this.loopJsonCache["microServicePolicies"]) {
				if (this.loopJsonCache["microServicePolicies"][policy]["name"] === name) {
					this.loopJsonCache["microServicePolicies"][policy]["pdpGroup"] = pdpGroup;
					this.loopJsonCache["microServicePolicies"][policy]["pdpSubgroup"] = pdpSubgroup;
				}
			}
	}

	updateGlobalProperties(newGlobalProperties) {
		this.loopJsonCache["globalPropertiesJson"] = newGlobalProperties;
	}

	updateOperationalPolicyProperties(name, newOpProperties) {
		for (var policy in this.loopJsonCache["operationalPolicies"]) {
				if (this.loopJsonCache["operationalPolicies"][policy]["name"] === name) {
					this.loopJsonCache["operationalPolicies"][policy]["configurationsJson"] = newOpProperties;
				}
			}
	}

	updateOperationalPolicyPdpGroup(name, pdpGroup, pdpSubgroup) {
		for (var policy in this.loopJsonCache["operationalPolicies"]) {
			if (this.loopJsonCache["operationalPolicies"][policy]["name"] === name) {
				this.loopJsonCache["operationalPolicies"][policy]["pdpGroup"] = pdpGroup;
				this.loopJsonCache["operationalPolicies"][policy]["pdpSubgroup"] = pdpSubgroup;
			}
		}
	}

	getLoopName() {
		return this.loopJsonCache["name"];
	}

	getOperationalPolicyJsonSchema() {
		return this.loopJsonCache["operationalPolicies"]["0"]["jsonRepresentation"];
	}

	getOperationalPolicies() {
		return this.loopJsonCache["operationalPolicies"];
	}

	getOperationalPoliciesNoJsonSchema() {
		var operationalPolicies = JSON.parse(JSON.stringify(this.loopJsonCache["operationalPolicies"]));
		delete operationalPolicies[0]["jsonRepresentation"];
		return operationalPolicies;
	}

	getGlobalProperties() {
		return this.loopJsonCache["globalPropertiesJson"];
	}

	getDcaeDeploymentProperties() {
		return this.loopJsonCache["globalPropertiesJson"]["dcaeDeployParameters"];
	}

	getMicroServicePolicies() {
		return this.loopJsonCache["microServicePolicies"];
	}

	getOperationalPolicyForName(name) {
		var opProperties=this.getOperationalPolicies();
		for (var policy in opProperties) {
			if (opProperties[policy]["name"] === name) {
				return opProperties[policy];
			}
		}
		return null;
	}

	getOperationalPolicyPropertiesForName(name) {
		var opConfig = this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
			return opConfig["configurationsJson"];
		}
		return null;
	}

	getOperationalPolicyJsonRepresentationForName(name) {
		var opConfig = this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
			return opConfig["jsonRepresentation"];
		}
		return null;
	}

	getOperationalPolicySupportedPdpGroup(name) {
		var opConfig=this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
		    if (opConfig["policyModel"]["policyPdpGroup"] !== undefined && opConfig["policyModel"]["policyPdpGroup"]["supportedPdpGroups"] !== undefined) {
			    return opConfig["policyModel"]["policyPdpGroup"]["supportedPdpGroups"];
			}
		}
		return [];
	}

	getOperationalPolicyPdpGroup(name) {
		var opConfig=this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
			return opConfig["pdpGroup"];
		}
		return null;
	}

	getOperationalPolicyPdpSubgroup(name) {
		var opConfig=this.getOperationalPolicyForName(name);
		if (opConfig !== null) {
			return opConfig["pdpSubgroup"];
		}
		return null;
	}

	getMicroServiceSupportedPdpGroup(name) {
		var microService=this.getMicroServiceForName(name);
		if (microService !== null) {
		    if (microService["policyModel"]["policyPdpGroup"] !== undefined && microService["policyModel"]["policyPdpGroup"]["supportedPdpGroups"] !== undefined) {
			    return microService["policyModel"]["policyPdpGroup"]["supportedPdpGroups"];
			}
		}
		return [];
	}

	getMicroServicePdpGroup(name) {
		var microService=this.getMicroServiceForName(name);
		if (microService !== null) {
			return microService["pdpGroup"];
		}
		return null;
	}

	getMicroServicePdpSubgroup(name) {
		var microService=this.getMicroServiceForName(name);
		if (microService !== null) {
			return microService["pdpSubgroup"];
		}
		return null;
	}

	getMicroServiceForName(name) {
		var msProperties=this.getMicroServicePolicies();
		for (var policy in msProperties) {
			if (msProperties[policy]["name"] === name) {
				return msProperties[policy];
			}
		}
		return null;
	}

	getMicroServicePropertiesForName(name) {
		var msConfig = this.getMicroServiceForName(name);
		if (msConfig !== null) {
			return msConfig["configurationsJson"];
		}
		return null;
	}

	getMicroServiceJsonRepresentationForName(name) {
		var msConfig = this.getMicroServiceForName(name);
		if (msConfig !== null) {
			return msConfig["jsonRepresentation"];
		}
		return null;
	}

	getResourceDetailsVfProperty() {
		return this.loopJsonCache["modelService"]["resourceDetails"]["VF"];
	}

	getResourceDetailsVfModuleProperty() {
		return this.loopJsonCache["modelService"]["resourceDetails"]["VFModule"];
	}

	getLoopLogsArray() {
		return this.loopJsonCache.loopLogs;
	}

	getComputedState() {
		return this.loopJsonCache.lastComputedState;
	}

	getComponentStates() {
		return this.loopJsonCache.components;
	}

	getTemplateName() {
	    if (this.getLoopTemplate() !== undefined) {
        			return this.getLoopTemplate().name;
        }
	    return null;
	}

	getLoopTemplate() {
		return this.loopJsonCache["loopTemplate"];
	}

	isOpenLoopTemplate() {
		var loopTemplate = this.getLoopTemplate();
		if(loopTemplate != null && loopTemplate["allowedLoopType"] === "OPEN") {
			return true;
		}
		return false;
	}

	getAllLoopElementModels() {
	    var loopTemplate = this.getLoopTemplate();
	    var loopElementModels = [];
	    if(loopTemplate != null) {
	        for (var element of loopTemplate['loopElementModelsUsed']) {
	            loopElementModels.push(element['loopElementModel'])
	        }
	    }
	    return loopElementModels;
	}
}
