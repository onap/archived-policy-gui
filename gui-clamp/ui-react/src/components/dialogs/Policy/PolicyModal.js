/*-
 * ============LICENSE_START=======================================================
 * ONAP POLICY-CLAMP
 * ================================================================================
 * Copyright (C) 2020-2021 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */

import React from 'react'
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';
import Select from 'react-select';
import Modal from 'react-bootstrap/Modal';
import styled from 'styled-components';
import LoopService from '../../../api/LoopService';
import LoopCache from '../../../api/LoopCache';
import { JSONEditor } from '@json-editor/json-editor/dist/jsoneditor.js';
import "@fortawesome/fontawesome-free/css/all.css"
import Alert from 'react-bootstrap/Alert';
import OnapConstant from '../../../utils/OnapConstants';
import OnapUtils from '../../../utils/OnapUtils';

const ModalStyled = styled(Modal)`
  background-color: transparent;
`

const DivWhiteSpaceStyled = styled.div`
  white-space: pre;
`

export default class PolicyModal extends React.Component {

  state = {
    show: true,
    loopCache: this.props.loopCache,
    jsonEditor: null,
    policyName: this.props.match.params.policyName,
    // This is to indicate whether it's an operational or config policy (in terms of loop instance)
    policyInstanceType: this.props.match.params.policyInstanceType,
    pdpGroup: null,
    pdpGroupList: [],
    pdpSubgroupList: [],
    chosenPdpGroup: '',
    chosenPdpSubgroup: '',
    showSucAlert: false,
    showFailAlert: false
  };

  constructor(props, context) {
    super(props, context);
    this.handleClose = this.handleClose.bind(this);
    this.handleSave = this.handleSave.bind(this);
    this.renderJsonEditor = this.renderJsonEditor.bind(this);
    this.handlePdpGroupChange = this.handlePdpGroupChange.bind(this);
    this.handlePdpSubgroupChange = this.handlePdpSubgroupChange.bind(this);
    this.createJsonEditor = this.createJsonEditor.bind(this);
    this.handleRefresh = this.handleRefresh.bind(this);
    this.disableAlert = this.disableAlert.bind(this);
    this.renderPdpGroupDropDown = this.renderPdpGroupDropDown.bind(this);
    this.renderOpenLoopMessage = this.renderOpenLoopMessage.bind(this);
    this.renderModalTitle = this.renderModalTitle.bind(this);
    this.readOnly = props.readOnly !== undefined ? props.readOnly : false;
  }

  handleSave() {
    var editorData = this.state.jsonEditor.getValue();
    var errors = this.state.jsonEditor.validate();
    errors = errors.concat(this.customValidation(editorData, this.state.loopCache.getTemplateName()));

    if (errors.length !== 0) {
      console.error("Errors detected during policy data validation ", errors);
      this.setState({
        showFailAlert: true,
        showMessage: 'Errors detected during policy data validation:\n' + OnapUtils.jsonEditorErrorFormatter(errors)
      });
      return;
    } else {
      console.info("NO validation errors found in policy data");
      if (this.state.policyInstanceType === OnapConstant.microServiceType) {
        this.state.loopCache.updateMicroServiceProperties(this.state.policyName, editorData);
        this.state.loopCache.updateMicroServicePdpGroup(this.state.policyName, this.state.chosenPdpGroup, this.state.chosenPdpSubgroup);
        LoopService.setMicroServiceProperties(this.state.loopCache.getLoopName(), this.state.loopCache.getMicroServiceForName(this.state.policyName)).then(resp => {
          this.setState({ show: false });
          this.props.history.push('/');
          this.props.loadLoopFunction(this.state.loopCache.getLoopName());
        });
      } else if (this.state.policyInstanceType === OnapConstant.operationalPolicyType) {
        this.state.loopCache.updateOperationalPolicyProperties(this.state.policyName, editorData);
        this.state.loopCache.updateOperationalPolicyPdpGroup(this.state.policyName, this.state.chosenPdpGroup, this.state.chosenPdpSubgroup);
        LoopService.setOperationalPolicyProperties(this.state.loopCache.getLoopName(), this.state.loopCache.getOperationalPolicies()).then(resp => {
          this.setState({ show: false });
          this.props.history.push('/');
          this.props.loadLoopFunction(this.state.loopCache.getLoopName());
        });
      }
    }
  }

  customValidation(editorData, templateName) {
    // method for sub-classes to override with customized validation
    return [];
  }

  handleClose() {
    this.setState({ show: false });
    this.props.history.push('/');
  }

  componentDidMount() {
    this.renderJsonEditor();
  }

  componentDidUpdate() {
    if (this.state.showSucAlert === true || this.state.showFailAlert === true) {
      let modalElement = document.getElementById("policyModal")
      if (modalElement) {
        modalElement.scrollTo(0, 0);
      }
    }
  }

  createJsonEditor(toscaModel, editorData) {
    /*JSONEditor.defaults.themes.myBootstrap4 = JSONEditor.defaults.themes.bootstrap4.extend({
            getTab: function(text,tabId) {
                var liel = document.createElement('li');
                liel.classList.add('nav-item');
                var ael = document.createElement("a");
                ael.classList.add("nav-link");
                ael.setAttribute("style",'padding:10px;max-width:160px;');
                ael.setAttribute("href", "#" + tabId);
                ael.setAttribute('data-toggle', 'tab');
                text.setAttribute("style",'word-wrap:break-word;');
                ael.appendChild(text);
                liel.appendChild(ael);
                return liel;
            }
        });*/
    return new JSONEditor(document.getElementById("editor"),
      {
        schema: toscaModel,
        startval: editorData,
        theme: 'bootstrap4',
        iconlib: 'fontawesome5',
        object_layout: 'grid',
        disable_properties: false,
        disable_edit_json: false,
        disable_array_reorder: true,
        disable_array_delete_last_row: true,
        disable_array_delete_all_rows: false,
        array_controls_top: true,
        keep_oneof_values: false,
        collapsed: true,
        show_errors: 'always',
        display_required_only: false,
        show_opt_in: false,
        prompt_before_delete: true,
        required_by_default: false
      })
  }

  renderJsonEditor() {
    console.debug("Rendering PolicyModal ", this.state.policyName);
    var toscaModel = {};
    var editorData = {};
    var pdpGroupValues = {};
    var chosenPdpGroupValue, chosenPdpSubgroupValue;
    if (this.state.policyInstanceType === OnapConstant.microServiceType) {
      toscaModel = this.state.loopCache.getMicroServiceJsonRepresentationForName(this.state.policyName);
      editorData = this.state.loopCache.getMicroServicePropertiesForName(this.state.policyName);
      pdpGroupValues = this.state.loopCache.getMicroServiceSupportedPdpGroup(this.state.policyName);
      chosenPdpGroupValue = this.state.loopCache.getMicroServicePdpGroup(this.state.policyName);
      chosenPdpSubgroupValue = this.state.loopCache.getMicroServicePdpSubgroup(this.state.policyName);
    } else if (this.state.policyInstanceType === OnapConstant.operationalPolicyType) {
      toscaModel = this.state.loopCache.getOperationalPolicyJsonRepresentationForName(this.state.policyName);
      editorData = this.state.loopCache.getOperationalPolicyPropertiesForName(this.state.policyName);
      pdpGroupValues = this.state.loopCache.getOperationalPolicySupportedPdpGroup(this.state.policyName);
      chosenPdpGroupValue = this.state.loopCache.getOperationalPolicyPdpGroup(this.state.policyName);
      chosenPdpSubgroupValue = this.state.loopCache.getOperationalPolicyPdpSubgroup(this.state.policyName);
    }

    if (toscaModel == null) {
      return;
    }

    var pdpSubgroupValues = [];
    if (typeof (chosenPdpGroupValue) !== "undefined") {
      var selectedPdpGroup = pdpGroupValues.filter(entry => (Object.keys(entry)[0] === chosenPdpGroupValue));
      pdpSubgroupValues = selectedPdpGroup[0][chosenPdpGroupValue].map((pdpSubgroup) => {
        return { label: pdpSubgroup, value: pdpSubgroup }
      });
    }
    this.setState({
      jsonEditor: this.createJsonEditor(toscaModel, editorData),
      pdpGroup: pdpGroupValues,
      pdpGroupList: pdpGroupValues.map(entry => {
        return { label: Object.keys(entry)[0], value: Object.keys(entry)[0] };
      }),
      pdpSubgroupList: pdpSubgroupValues,
      chosenPdpGroup: chosenPdpGroupValue,
      chosenPdpSubgroup: chosenPdpSubgroupValue
    })
  }

  handlePdpGroupChange(e) {
    var selectedPdpGroup = this.state.pdpGroup.filter(entry => (Object.keys(entry)[0] === e.value));
    const pdpSubgroupValues = selectedPdpGroup[0][e.value].map((pdpSubgroup) => {
      return { label: pdpSubgroup, value: pdpSubgroup }
    });
    if (this.state.chosenPdpGroup !== e.value) {
      this.setState({
        chosenPdpGroup: e.value,
        chosenPdpSubgroup: '',
        pdpSubgroupList: pdpSubgroupValues
      });
    }
  }

  handlePdpSubgroupChange(e) {
    this.setState({ chosenPdpSubgroup: e.value });
  }

  handleRefresh() {
    var newLoopCache, toscaModel, editorData;
    if (this.state.policyInstanceType === OnapConstant.microServiceType) {
      LoopService.refreshMicroServicePolicyJson(this.state.loopCache.getLoopName(), this.state.policyName).then(data => {
        newLoopCache = new LoopCache(data);
        toscaModel = newLoopCache.getMicroServiceJsonRepresentationForName(this.state.policyName);
        editorData = newLoopCache.getMicroServicePropertiesForName(this.state.policyName);
        document.getElementById("editor").innerHTML = "";
        this.setState({
          loopCache: newLoopCache,
          jsonEditor: this.createJsonEditor(toscaModel, editorData),
          showSucAlert: true,
          showMessage: "Successfully refreshed"
        });
      })
        .catch(error => {
          console.error("Error while refreshing the Operational Policy Json Representation");
          this.setState({
            showFailAlert: true,
            showMessage: "Refreshing of UI failed"
          });
        });
    } else if (this.state.policyInstanceType === OnapConstant.operationalPolicyType) {
      LoopService.refreshOperationalPolicyJson(this.state.loopCache.getLoopName(), this.state.policyName).then(data => {
        var newLoopCache = new LoopCache(data);
        toscaModel = newLoopCache.getOperationalPolicyJsonRepresentationForName(this.state.policyName);
        editorData = newLoopCache.getOperationalPolicyPropertiesForName(this.state.policyName);
        document.getElementById("editor").innerHTML = "";
        this.setState({
          loopCache: newLoopCache,
          jsonEditor: this.createJsonEditor(toscaModel, editorData),
          showSucAlert: true,
          showMessage: "Successfully refreshed"
        });
      })
        .catch(error => {
          console.error("Error while refreshing the Operational Policy Json Representation");
          this.setState({
            showFailAlert: true,
            showMessage: "Refreshing of UI failed"
          });
        });
    }
  }

  disableAlert() {
    this.setState({ showSucAlert: false, showFailAlert: false });
  }

  renderPdpGroupDropDown() {
    if (this.state.policyInstanceType !== OnapConstant.operationalPolicyType || !this.state.loopCache.isOpenLoopTemplate()) {
      return (
        <Form.Group as={ Row } controlId="formPlaintextEmail">
          <Form.Label column sm="2">Pdp Group Info</Form.Label>
          <Col sm="3">
            <Select value={ { label: this.state.chosenPdpGroup, value: this.state.chosenPdpGroup } } onChange={ this.handlePdpGroupChange } options={ this.state.pdpGroupList }/>
          </Col>
          <Col sm="3">
            <Select value={ { label: this.state.chosenPdpSubgroup, value: this.state.chosenPdpSubgroup } } onChange={ this.handlePdpSubgroupChange } options={ this.state.pdpSubgroupList }/>
          </Col>
        </Form.Group>
      );
    }
  }

  renderOpenLoopMessage() {
    if (this.state.policyInstanceType === OnapConstant.operationalPolicyType && this.state.loopCache.isOpenLoopTemplate()) {
      return (
        "Operational Policy cannot be configured as only Open Loop is supported for this Template!"
      );
    }
  }

  renderModalTitle() {
    return (
      <Modal.Title>Edit the policy</Modal.Title>
    );
  }

  renderButton() {
    var allElement = [(<Button variant="secondary" onClick={ this.handleClose }>
      Close
    </Button>)];
    if (this.state.policyInstanceType !== OnapConstant.operationalPolicyType || !this.state.loopCache.isOpenLoopTemplate()) {
      allElement.push((
        <Button variant="primary" disabled={ this.readOnly } onClick={ this.handleSave }>
          Save Changes
        </Button>
      ));
      allElement.push((
        <Button variant="primary" disabled={ this.readOnly } onClick={ this.handleRefresh }>
          Refresh
        </Button>
      ));
    }
    return allElement;
  }

  render() {
    return (
      <ModalStyled size="xl" backdrop="static" keyboard={ false } show={ this.state.show } onHide={ this.handleClose }>
        <Modal.Header closeButton>
          { this.renderModalTitle() }
        </Modal.Header>
        <Alert variant="success" show={ this.state.showSucAlert } onClose={ this.disableAlert } dismissible>
          <DivWhiteSpaceStyled>
            { this.state.showMessage }
          </DivWhiteSpaceStyled>
        </Alert>
        <Alert variant="danger" show={ this.state.showFailAlert } onClose={ this.disableAlert } dismissible>
          <DivWhiteSpaceStyled>
            { this.state.showMessage }
          </DivWhiteSpaceStyled>
        </Alert>
        <Modal.Body>
          { this.renderOpenLoopMessage() }
          <div id="editor"/>
          { this.renderPdpGroupDropDown() }
        </Modal.Body>
        <Modal.Footer>
          { this.renderButton() }
        </Modal.Footer>
      </ModalStyled>
    );
  }
}
