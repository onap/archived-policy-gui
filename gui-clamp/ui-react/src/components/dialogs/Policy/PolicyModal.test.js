/*-
 * ============LICENSE_START=======================================================
 * ONAP CLAMP
 * ================================================================================
 * Copyright (C) 2020 AT&T Intellectual Property. All rights
 *                             reserved.
 * ================================================================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ============LICENSE_END============================================
 * ===================================================================
 *
 */
import React from 'react';
import { mount } from 'enzyme';
import PolicyModal from './PolicyModal';
import LoopCache from '../../../api/LoopCache';
import LoopService from '../../../api/LoopService';
import OnapConstant from '../../../utils/OnapConstants';

describe('Verify PolicyModal', () => {
    beforeEach(() => {
        fetch.resetMocks();
        fetch.mockImplementation(() => {
            return Promise.resolve({
                ok: true,
                status: 200,
                text: () => "OK"
            });
        });
    })
    const loopCacheStr = {
            "name": "LOOP_Jbv1z_v1_0_ResourceInstanceName1_tca",
            "operationalPolicies": [{
                "name": "OPERATIONAL_h2NMX_v1_0_ResourceInstanceName1_tca",
                "configurationsJson": {
                    "operational_policy": {
                        "controlLoop": {},
                        "policies": []
                    }
                },
                "policyModel": {"policyPdpGroup": {"supportedPdpGroups":[{"monitoring": ["xacml"]}]}},
                "jsonRepresentation" : {"schema": {}}
             }]
    };
    const loopCache = new LoopCache(loopCacheStr);
    const historyMock = { push: jest.fn() };
    const flushPromises = () => new Promise(setImmediate);
    const match = {params: {policyName:"OPERATIONAL_h2NMX_v1_0_ResourceInstanceName1_tca", policyInstanceType: OnapConstant.operationalPolicyType}}
        
    it('Test handleClose', () => {
      const handleClose = jest.spyOn(PolicyModal.prototype,'handleClose');
      const component = mount(<PolicyModal history={historyMock} match={match} loopCache={loopCache}/>)

      component.find('[variant="secondary"]').prop('onClick')();

      expect(handleClose).toHaveBeenCalledTimes(1);
      expect(component.state('show')).toEqual(false);
      expect(historyMock.push.mock.calls[0]).toEqual([ '/']);
    });

    it('Test handleSave', async () => {
        const loadLoopFunction = jest.fn();
        const handleSave = jest.spyOn(PolicyModal.prototype,'handleSave');
        const component = mount(<PolicyModal history={historyMock} 
                          loopCache={loopCache} match={match} loadLoopFunction={loadLoopFunction} />)

        component.find('[variant="primary"]').get(0).props.onClick();
        await flushPromises();
        component.update();

        expect(handleSave).toHaveBeenCalledTimes(1);
        expect(component.state('show')).toEqual(false);
        expect(historyMock.push.mock.calls[0]).toEqual([ '/']);
    });

    it('Test handleRefresh', async () => {
        LoopService.refreshOperationalPolicyJson = jest.fn().mockImplementation(() => {
            return Promise.resolve(loopCacheStr);
        });
        const updateLoopFunction = jest.fn();
        const handleRefresh = jest.spyOn(PolicyModal.prototype,'handleRefresh');
        const component = mount(<PolicyModal loopCache={loopCache} match={match} updateLoopFunction={updateLoopFunction} />)

        component.find('[variant="primary"]').get(1).props.onClick();
        await flushPromises();
        component.update();

        expect(handleRefresh).toHaveBeenCalledTimes(1);
        expect(component.state('show')).toEqual(true);
        expect(component.state('showSucAlert')).toEqual(true);
        expect(component.state('showMessage')).toEqual("Successfully refreshed");
    });

    it('Test handlePdpGroupChange', () => {
        const component = mount(<PolicyModal loopCache={loopCache} match={match} />)
        component.setState({ 
                "pdpGroup": [{"option1":["subPdp1","subPdp2"]}],
                "chosenPdpGroup": "option2"
        });
        expect(component.state('chosenPdpGroup')).toEqual("option2");

        const instance = component.instance();
        const event = {label:"option1", value:"option1"}
        instance.handlePdpGroupChange(event);
        expect(component.state('chosenPdpGroup')).toEqual("option1");
        expect(component.state('chosenPdpSubgroup')).toEqual("");
        expect(component.state('pdpSubgroupList')).toEqual([{label:"subPdp1", value:"subPdp1"}, {label:"subPdp2", value:"subPdp2"}]);
    });

    it('Test handlePdpSubgroupChange', () => {
        const component = mount(<PolicyModal loopCache={loopCache} match={match} />)

        const instance = component.instance();
        const event = {label:"option1", value:"option1"}
        instance.handlePdpSubgroupChange(event);
        expect(component.state('chosenPdpSubgroup')).toEqual("option1");
    });
});