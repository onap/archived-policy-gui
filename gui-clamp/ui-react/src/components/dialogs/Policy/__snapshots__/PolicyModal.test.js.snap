// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify PolicyModal Test the render method 1`] = `
<Styled(Modal)
  backdrop="static"
  keyboard={false}
  onHide={[Function]}
  show={true}
  size="xl"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle>
      Edit the policy
    </ModalTitle>
  </ModalHeader>
  <Alert
    closeLabel="Close alert"
    dismissible={true}
    onClose={[Function]}
    show={false}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "render": [Function],
      }
    }
    variant="success"
  >
    <styled.div />
  </Alert>
  <Alert
    closeLabel="Close alert"
    dismissible={true}
    onClose={[Function]}
    show={false}
    transition={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "appear": false,
          "in": false,
          "mountOnEnter": false,
          "timeout": 300,
          "unmountOnExit": false,
        },
        "render": [Function],
      }
    }
    variant="danger"
  >
    <styled.div />
  </Alert>
  <ModalBody>
    <div
      id="editor"
    />
    <FormGroup
      as={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "noGutters": false,
          },
          "render": [Function],
        }
      }
      controlId="formPlaintextEmail"
    >
      <FormLabel
        column={true}
        sm="2"
        srOnly={false}
      >
        Pdp Group Info
      </FormLabel>
      <Col
        sm="3"
      >
        <StateManager
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          onChange={[Function]}
          options={
            Array [
              Object {
                "label": "monitoring",
                "value": "monitoring",
              },
            ]
          }
          value={
            Object {
              "label": undefined,
              "value": undefined,
            }
          }
        />
      </Col>
      <Col
        sm="3"
      >
        <StateManager
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          onChange={[Function]}
          options={Array []}
          value={
            Object {
              "label": undefined,
              "value": undefined,
            }
          }
        />
      </Col>
    </FormGroup>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      disabled={false}
      key="close"
      onClick={[Function]}
      variant="secondary"
    >
      Close
    </Button>
    <Button
      active={false}
      disabled={false}
      key="save"
      onClick={[Function]}
      variant="primary"
    >
      Save Changes
    </Button>
    <Button
      active={false}
      disabled={false}
      key="refresh"
      onClick={[Function]}
      variant="primary"
    >
      Refresh
    </Button>
  </ModalFooter>
</Styled(Modal)>
`;
